### DISCUSSION TITLE
Repeater with data through one-to-many relationship?

### DISCUSSION THREAD

**Discussion by yaplej** [filamentphp/filament#2645] [Help] [answered]
Hello,

I create a HasManyRepeater that uses a relationship to another Model.  How would I create a Form that to create the record for the 2nd relationship?

This displays fine but will not save because it does not create an `approval` record to populate `approval_id` in the `changerequestapproval` relationship.

```
Forms\Components\HasManyRepeater::make('changerequestapproval')
                                    ->label('Approvals')
                                    ->relationship('changerequestapproval')
                                    ->schema([
                                        Forms\Components\Select::make('user_id')
                                            ->label('Name')
                                            ->options(User::all()->pluck('name', 'id'))
                                            ->searchable()
                                            ->required(),
                                        Forms\Components\Radio::make('approved')
                                            ->options([
                                                '1' => 'Approve',
                                                '0' => 'Decline',
                                                'null' => 'Abstain'
                                            ])
                                            ->inline('true'),
                                    ])

```

The Model for the HasManyRepeater
```
    protected $table = 'change_request_approvals';

    /**
     * The attributes that are mass assignable.
     *
     * @var array<int, string>
     */
    protected $fillable = [
        'change_request_id',
        'approval_id',
    ];

    public function change_request()
    {
        return $this->belongsTo(ChangeRequest::class);
    }

    public function approval()
    {
        return $this->belongsTo(Approval::class);
    }

}
```

**Comment by danharrin** [comment]
As discussed on Discord, define a HasManyThrough relationship, then try using this with RelationshipRepeater. If that works, we can add a dedicated HasManyThroughRepeater.

**Reply by yaplej** [reply]
Thank you for the follow-up.  

I just figured out I was using the wrong relationship type in Laravel.  HasManyThrough appears to be suited for one-to-one-to-many relationships but I have my database designed for one-to-many-to-one or one-to-many-to-many.  For that I should have been using belongsToMany.  It works great in the RelationshipManager now but still does not work with the Repeater. 

```
Filament\Forms\Components\RelationshipRepeater::getRelationship(): Return value must be of type Illuminate\Database\Eloquent\Relations\HasOneOrMany, Illuminate\Database\Eloquent\Relations\BelongsToMany returned
``` 

I had to make sure things worked with the RelationshipManager before trying to customize anything.  I think the next step is the same though.  Making the Repeater work with BelongsToMany.

**Comment by yaplej** [comment, marked as answer]
My data required that the RelationshipRepeater to work with BelongsToMany relationships.  I updated the RelationshipRepeater so it works with that type of relationship and everything is working how I wanted now.
https://github.com/laravel-filament/filament/pull/2738
